cajita/src/Cajita_Halo.hpp

```diff
@@ -427,37 +503,488 @@ class Halo
             // Otherwise unpack the next buffer and apply the reduce operation.
             else
             {
-                auto subview =
-                    createSubview( view, _owned_spaces[unpack_index] );
-                auto owned_buffer = _owned_buffers[unpack_index];
-                IndexSpace<4> scatter_space(
-                    {static_cast<long>( subview.extent( 0 ) ),
-                     static_cast<long>( subview.extent( 1 ) ),
-                     static_cast<long>( subview.extent( 2 ) ),
-                     static_cast<long>( subview.extent( 3 ) )} );
-                Kokkos::parallel_for(
-                    "Cajita::Halo::scatterOwned",
-                    createExecutionPolicy( scatter_space, execution_space() ),
-                    KOKKOS_LAMBDA( const int i, const int j, const int k,
-                                   const int l ) {
-                        reduce( subview, owned_buffer, i, j, k, l );
-                    } );
+                unpackBuffer( reduce_op, exec_space,
+                              _owned_buffers[unpack_index],
+                              _owned_steering[unpack_index], arrays.view()... );
             }
 
             // Wait on send requests.
             MPI_Waitall( num_n, requests.data() + num_n, MPI_STATUSES_IGNORE );
         }
     }
 
+  public:
+    // Get the byte sizes of the array value types.
+    template <class Array_t>
+    void getByteSizes( std::vector<int> &byte_sizes, const int array_idx,
+                       const Array_t & )
+    {
+        byte_sizes[array_idx] = sizeof( typename Array_t::value_type );
+    }
+
+    template <class Array_t, class... ArrayTypes>
+    void getByteSizes( std::vector<int> &byte_sizes, const int array_idx,
+                       const Array_t &, ArrayTypes... arrays )
+    {
+        byte_sizes[array_idx] = sizeof( typename Array_t::value_type );
+        getByteSizes( byte_sizes, array_idx + 1, arrays... );
+    }
```

Comment:
You know the size of `byte_sizes` at compile time so it could/should be a `std::array`.
Also consider returning it instead of taking it by reference to improve readability.

Reply:
I have changed this to `std::array`